filecoin-proofs/src/fr32_reader.rs

```diff
@@ -1,157 +1,105 @@
+use byte_slice_cast::*;
 use std::io;
 
-const DATA_BITS: u64 = 254;
-const TARGET_BITS: u64 = 256;
-
-#[derive(Debug)]
 pub struct Fr32Reader<R> {
     /// The source being padded.
     source: R,
-    /// How much of the target already was `read` from, in bits.
-    target_offset: u64,
-    /// Currently read byte.
-    buffer: Buffer,
+    /// Currently read block.
+    in_buffer: [u8; NUM_U128S_PER_BLOCK * 16],
+    /// Currently writing out block.
+    out_buffer: [u128; NUM_U128S_PER_BLOCK],
+    /// How many blocks are left in the buffers.
+    to_process: usize,
+    /// How many bytes of the out_buffer are already read out.
+    out_offset: usize,
     /// Are we done reading?
     done: bool,
 }
 
+const NUM_U128S_PER_BLOCK: usize = 8;
+const MASK_SKIP_HIGH_2: u128 = 0b0011_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111;
+
```

Comment:
I find that much harder to inspect to be honest

Reply:
Really? It's easy for you to see at a glance that this is wrong?

```
0b0011_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111
```

Reply:
neither of them is "easy" for me, but in thase case I simply count the 1s and know if it is good, but your notation requires me to think through bitshfiting rules, and how the subtraction operator is used on a bit level.

Reply:
My main point was to avoid counting the `1`s — not the specific construction used to avoid it. Especially with an appropriate comment, I'm not averse to asking readers to think through bitshifting rules.

A simpler version which requires less thinking and still no counting would be:

```
const MASK_SKIP_HIGH_2: u128 =  u128::MAX >> 2;
```

Maybe that is the best of both worlds?